src/image/ImageBaseFitsReader.cc

```diff
@@ -0,0 +1,209 @@
+ * This product includes software developed by the LSST Project
+ * (https://www.lsst.org).
+ * See the COPYRIGHT file at the top-level directory of this distribution
+ * for details of code ownership.
+ *
+ * This program is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License as published by
+ * the Free Software Foundation, either version 3 of the License, or
+ * (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program.  If not, see <https://www.gnu.org/licenses/>.
+ */
+
+#include "lsst/afw/image/ImageBaseFitsReader.h"
+#include "lsst/afw/geom/wcsUtils.h"
+
+namespace lsst { namespace afw { namespace image {
+
+ImageBaseFitsReader::ImageBaseFitsReader(std::string const& fileName, int hdu) :
+    _ownsFitsFile(true),
+    _hdu(0),
+    _fitsFile(new fits::Fits(fileName, "r", fits::Fits::AUTO_CLOSE | fits::Fits::AUTO_CHECK))
+{
+    _fitsFile->setHdu(hdu);
+    _fitsFile->checkCompressedImagePhu();
+    _hdu = _fitsFile->getHdu();
+}
+
+ImageBaseFitsReader::ImageBaseFitsReader(fits::MemFileManager& manager, int hdu) :
+    _ownsFitsFile(true),
+    _hdu(0),
+    _fitsFile(new fits::Fits(manager, "r", fits::Fits::AUTO_CLOSE | fits::Fits::AUTO_CHECK))
+{
+    _fitsFile->setHdu(hdu);
+    _fitsFile->checkCompressedImagePhu();
+    _hdu = _fitsFile->getHdu();
+}
+
+ImageBaseFitsReader::ImageBaseFitsReader(fits::Fits * fitsFile) :
+    _ownsFitsFile(false),
+    _hdu(0),
+    _fitsFile(fitsFile)
+{
+    if (_fitsFile) {
+        if (_fitsFile->getHdu() == 0 && _fitsFile->getImageDim() == 0) {
+            _fitsFile->setHdu(fits::DEFAULT_HDU);
+        }
+        _fitsFile->checkCompressedImagePhu();
+        _hdu = _fitsFile->getHdu();
+    }
+}
+
+ImageBaseFitsReader::~ImageBaseFitsReader() noexcept {
+    if (_ownsFitsFile) {
+        delete _fitsFile;
+    }
+}
+
+namespace {
+
+void checkFitsFile(fits::Fits * f) {
+    if (f == nullptr) {
+        throw LSST_EXCEPT(
+            fits::FitsError,
+            "FitsReader not initialized; desired HDU is probably missing."
+        );
+    }
+}
+
+} // anonymous
+
+std::string ImageBaseFitsReader::readDType() const {
+    checkFitsFile(_fitsFile);
+    fits::HduMoveGuard guard(*_fitsFile, _hdu);
+    return _fitsFile->getImageDType();
+}
+
+lsst::geom::Box2I ImageBaseFitsReader::readBBox(ImageOrigin origin) {
+    readMetadata();  // guarantees _bbox is initialized
+    if (origin == LOCAL) {
+        return lsst::geom::Box2I(lsst::geom::Point2I(), _bbox.getDimensions());
+    }
+    return _bbox;
+}
+
+lsst::geom::Point2I ImageBaseFitsReader::readXY0(lsst::geom::Box2I const & bbox, ImageOrigin origin) {
```

Comment:
The `origin` argument here actually only affects how the `bbox` argument is interpreted, so it can't affect the result when `bbox` is empty.  I'll document that and raise an exception if we ever get `LOCAL` with an empty `bbox`, as that probably means someone is confused about what that does.

Reply:
Update: raising an exception is out, because there existing code expecting to be able to pass `LOCAL` (and have it ignored).  But the fact that `origin` is ignored if `bbox` is empty is now documented.